File: 2015/2/python/soln.py
from pathlib import Path
import math

puzzle_input = Path("./input").read_text().strip()


def surface_area(width, length, height):
    sides = [length * width, width * height, height * length]
    # Add margin of error: size of smallest side
    return sum(sides) * 2 + min(sides)


def ribbon_length(dimensions):
    bow = math.prod(dimensions)
    dimensions.remove(max(dimensions))
    return 2 * sum(dimensions) + bow


def part1(puzzle):
    total_sq_feet = 0
    for line in puzzle.splitlines():
        dimensions = [int(i) for i in line.split("x")]
        total_sq_feet += surface_area(*dimensions)
    return total_sq_feet


def part2(puzzle):
    total_ribbon = 0
    for line in puzzle.splitlines():
        dimensions = [int(i) for i in line.split("x")]
        total_ribbon += ribbon_length(dimensions)
    return total_ribbon


assert surface_area(2, 3, 4) == 58
assert surface_area(1, 1, 10) == 43
assert ribbon_length([2, 3, 4]) == 34
assert ribbon_length([1, 1, 10]) == 14

result = part1(puzzle_input)
print(f"Part 1 {result} square feet")
result = part2(puzzle_input)
print(f"Part 2 {result} feet of ribbon")

File: 2015/2/julia/soln.jl
using Printf

puzzle_input = open(joinpath(dirname(Base.PROGRAM_FILE), "../input")) do file
    read(file, String)
end


function surface_area(width, length, height)
    sides = [length * width, width * height, height * length]
    # Add margin of error: size of smallest side
    return sum(sides) * 2 + min(sides...)
end

function ribbon_length(dimensions)
    bow = prod(dimensions)
    highest = max(dimensions...)
    deleteat!(dimensions, findfirst(x->x==highest, dimensions))
    return 2 * sum(dimensions) + bow
end

function part1(data)
    total_sq_feet = 0
    for line in split(data, "\n")
        x, y, z = [parse(Int64, i) for i in split(line, "x")]
        total_sq_feet += surface_area(x, y, z)
    end
    return total_sq_feet
end


function part2(data)
    total_ribbon = 0
    for line in split(data, "\n")
        dimensions = [parse(Int64, i) for i in split(line, "x")]
        total_ribbon += ribbon_length(dimensions)
    end
    return total_ribbon
end

@printf "Part 1: %s\n" part1(puzzle_input)
@printf "Part 2: %s\n" part2(puzzle_input)
